signer/src/stacks/contracts.rs

```diff
@@ -987,11 +1070,118 @@ impl AsContractCall for RejectWithdrawalV1 {
+                // fails #2
+                return Err(
+                    WithdrawalRejectErrorMsg::WithdrawalRequestFulfilled.into_error(req_ctx, self)
+                );
+            }
+            WithdrawalRequestStatus::Unconfirmed => {
+                // fails #1
+                return Err(WithdrawalRejectErrorMsg::WithdrawalRequestUnconfirmed
+                    .into_error(req_ctx, self));
+            }
+        }
+
+        // 3. The request is final. Check that we've seen six new bitcoin blocks since observing
+        //    the bitcoin anchor block associated with the Stacks block confirming the withdrawal
+        //    request transaction. Fail the withdrawal request if we’ve observed less than 6
+        //    bitcoin blocks since the anchor block.
+
+        let request_block_height = report.block_height;
+
+        let Some(tip_block_hash) = ctx.get_storage().get_bitcoin_canonical_chain_tip().await?
+        else {
+            return Err(WithdrawalRejectErrorMsg::NoBitcoinChainTip.into_error(req_ctx, self));
+        };
+
+        let Some(tip_bitcoin_block) = ctx.get_storage().get_bitcoin_block(&tip_block_hash).await?
+        else {
+            return Err(WithdrawalRejectErrorMsg::MissingBitcoinBlock.into_error(req_ctx, self));
+        };
+
+        let tip_block_height = tip_bitcoin_block.block_height;
+
+        if tip_block_height < (request_block_height + 6) {
+            return Err(
+                WithdrawalRejectErrorMsg::WithdrawalRequestNotFinal.into_error(req_ctx, self)
+            );
+        }
+
+        // 4. The request is collectively rejected or expired.
+        let signer_votes = ctx
+            .get_storage()
+            .get_withdrawal_request_signer_votes(&self.id, &req_ctx.aggregate_key)
+            .await
+            .map_err(|_| WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self))?;
+        let signer_bitmap = BitArray::<[u8; 16]>::from(signer_votes);
+
+        if signer_bitmap != self.signer_bitmap {
+            return Err(WithdrawalRejectErrorMsg::BitmapMismatch.into_error(req_ctx, self));
+        }
+
+        let withdrawal_signers = &ctx
+            .get_storage()
+            .get_withdrawal_signers(self.id.request_id, &self.id.block_hash)
+            .await
+            .map_err(|_| WithdrawalRejectErrorMsg::RequestMissing.into_error(req_ctx, self))?;
+
+        let mut rejected_count = 0;
+        for signer in withdrawal_signers {
+            if !signer.is_accepted {
+                rejected_count += 1;
+            }
+        }
+
+        let threshold = req_ctx.signatures_required;
+        let num_signers = ctx.state().current_signer_set().get_signers().len();
+        let reject_threshold = num_signers - threshold as usize + 1;
+
+        if rejected_count < reject_threshold {
+            return Err(
+                WithdrawalRejectErrorMsg::WithdrawalRequestNotRejected.into_error(req_ctx, self)
+            );
+        }
```

Comment:
This part is tricky since there are two reasons for why we would reject a withdrawal.
1. The withdrawal request has enough votes against.
2. The withdrawal request has expired. This happens after `WITHDRAWAL_BLOCKS_EXPIRY` blocks (the one from my PR [here](https://github.com/stacks-network/sbtc/pull/1352/files#diff-6cd449eb3c30449f70c0d11614a6b525227e18691745335775bf41baec8d7482R121-R126)) from the height of the `bitcoin_block_height` in the `WithdrawalRequestReport` (it's `block_height` on main but we're renaming it).

So we need to check for either condition here.

Reply:
Done

Reply:
I don't think that this one is done.

Reply:
Oh sorry I hadn't pushed when I commented.  Now it's:

https://github.com/stacks-network/sbtc/blob/478-validate-reject-withdrawal-request/signer/src/stacks/contracts.rs#L1162

Reply:
Looking back I don't think this check is sufficient.

Reply:
The logic is that if we negate either of these two conditions, that is sufficient for a valid rejection.  It's equivalent to:

    if !(blocks_observed >= WITHDRAWAL_BLOCKS_EXPIRY || rejected_count >= reject_threshold) {

But what if *both* conditions are negated?  Is it still a valid rejection if we have a threshold but it's expired?  I don't think so.

Maybe what we need is

    if !(blocks_observed >= WITHDRAWAL_BLOCKS_EXPIRY ^ rejected_count >= reject_threshold) {

Reply:
Ok I split the expired test into two: one with a threshold of rejection votes and one without.  Both tests pass.  We already test the case where we aren't expired but we are rejected (that's happy path).  But now we also need a not expired AND not rejected test.  Adding that now.

Reply:
Ok test added.  I think we're good now?

Reply:
After thinking about this some more, we I propose that we only care about whether the request is expired.

Reply:
That's interesting.  Certainly if we are expired, we should fail validation regardless of how many rejection votes we've received.  But if we aren't expired, but don't have a threshold of rejection votes, shouldn't the rejection request fail validation?

I'll wait to see your logic in your ticket.

Reply:
The ticket is here https://github.com/stacks-network/sbtc/issues/1363.

Basically, I think that we should fail validation if the `reject-withdrawal-request` contract call if the withdrawal request has not expired, regardless of the votes against.

Alternatively, we might want to allow for validation of the `reject-withdrawal-request` contract call if we receive a consensus number of votes against or something. But we'd also need to modify how we bitcoin validation as well, but we need to do that regardless.

Reply:
Done.